File: android/src/main/java/com/photovideoeditor/photo/render/PhotoLayerRenderer.kt
package com.photovideoeditor.photo.render

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.text.TextPaint

/**
 * Burns the layer stack onto a bitmap. Used identically by the live preview
 * and the exporter (different bitmap resolutions, same code and normalized
 * coordinates) so what you see matches what gets exported.
 */
object PhotoLayerRenderer {
  private val builtinStickers = mapOf(
    "smile" to "😀",
    "heart" to "❤️",
    "star" to "⭐",
    "fire" to "🔥",
    "thumbsUp" to "👍",
    "sun" to "☀️",
    "clap" to "👏",
    "sparkles" to "✨"
  )

  fun builtinStickerIds(): List<String> = builtinStickers.keys.toList()
  fun glyphFor(stickerId: String): String = builtinStickers[stickerId] ?: "★"

  fun render(base: Bitmap, layers: List<PhotoLayer>, bitmapResolver: (String) -> Bitmap?): Bitmap {
    if (layers.isEmpty()) return base
    val result = base.copy(Bitmap.Config.ARGB_8888, true)
    val canvas = Canvas(result)
    layers.filter { it.visible }.forEach { layer ->
      val geometry = OverlayGeometry.map(layer, result.width, result.height)
      canvas.save()
      canvas.translate(geometry.centerX, geometry.centerY)
      canvas.rotate(geometry.rotationDegrees)
      canvas.scale(geometry.scale, geometry.scale)
      val alpha = (geometry.opacity * 255).toInt()
      when (layer.type) {
        LayerType.TEXT -> drawText(canvas, layer, geometry.shortSide, alpha)
        LayerType.STICKER -> drawSticker(canvas, layer, geometry.shortSide, alpha, bitmapResolver)
        LayerType.OVERLAY -> drawOverlay(canvas, layer, geometry.shortSide, alpha, bitmapResolver)
        LayerType.DRAWING -> drawStroke(canvas, layer, result.width.toFloat(), result.height.toFloat(), geometry.shortSide, alpha)
      }
      canvas.restore()
    }
    return result
  }

  private fun drawText(canvas: Canvas, layer: PhotoLayer, shortSide: Float, alpha: Int) {
    val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
      color = layer.textColor
      this.alpha = alpha
      textSize = layer.fontSize / 48f * (shortSide * 0.06f)
      textAlign = Paint.Align.CENTER
      layer.fontFamily?.let { family ->
        try {
          typeface = android.graphics.Typeface.create(family, android.graphics.Typeface.NORMAL)
        } catch (_: Exception) {
          // Fall back to the default typeface if the family cannot be resolved.
        }
      }
    }
    val lines = layer.text.ifBlank { " " }.split("\n")
    val lineHeight = paint.fontSpacing
    val totalHeight = lineHeight * lines.size
    lines.forEachIndexed { index, line ->
      canvas.drawText(line, 0f, -totalHeight / 2f + lineHeight * (index + 1) - paint.descent(), paint)
    }
  }

  private fun drawSticker(canvas: Canvas, layer: PhotoLayer, shortSide: Float, alpha: Int, resolver: (String) -> Bitmap?) {
    val size = shortSide * 0.18f
    val uri = layer.stickerUri
    if (uri != null) {
      resolver(uri)?.let { bitmap ->
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { this.alpha = alpha }
        canvas.drawBitmap(bitmap, null, RectF(-size / 2, -size / 2, size / 2, size / 2), paint)
        return
      }
    }
    val glyph = layer.stickerId?.let { glyphFor(it) } ?: "★"
    val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
      this.alpha = alpha
      textSize = size
      textAlign = Paint.Align.CENTER
    }
    canvas.drawText(glyph, 0f, size / 2f - paint.descent(), paint)
  }

  private fun drawOverlay(canvas: Canvas, layer: PhotoLayer, shortSide: Float, alpha: Int, resolver: (String) -> Bitmap?) {
    val uri = layer.overlayUri ?: return
    val bitmap = resolver(uri) ?: return
    val aspect = layer.overlayAspectRatio ?: (bitmap.width.toFloat() / bitmap.height.toFloat())
    val baseSize = shortSide * 0.6f
    val width: Float
    val height: Float
    if (aspect >= 1f) {
      width = baseSize
      height = baseSize / aspect
    } else {
      height = baseSize
      width = baseSize * aspect
    }
    val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { this.alpha = alpha }
    canvas.drawBitmap(bitmap, null, RectF(-width / 2, -height / 2, width / 2, height / 2), paint)
  }

  private fun drawStroke(canvas: Canvas, layer: PhotoLayer, width: Float, height: Float, shortSide: Float, alpha: Int) {
    if (layer.drawPoints.size < 2) return
    val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
      color = layer.drawColor
      this.alpha = alpha
      style = Paint.Style.STROKE
      strokeWidth = layer.drawStrokeWidth * shortSide
      strokeCap = Paint.Cap.ROUND
      strokeJoin = Paint.Join.ROUND
    }
    val path = Path()
    layer.drawPoints.forEachIndexed { index, (dx, dy) ->
      val px = dx * width
      val py = dy * height
      if (index == 0) path.moveTo(px, py) else path.lineTo(px, py)
    }
    canvas.drawPath(path, paint)
  }
}

File: ios/Photo/Rendering/PhotoLayerRenderer.swift
import UIKit

/// Burns the layer stack onto a `UIImage`. Used identically by the live
/// preview and the exporter (different image resolutions, same code and
/// normalized coordinates) so what you see matches what gets exported.
/// Mirrors `PhotoLayerRenderer.kt`.
enum PhotoLayerRenderer {
  private static let builtinStickers: [String: String] = [
    "smile": "😀", "heart": "❤️", "star": "⭐", "fire": "🔥",
    "thumbsUp": "👍", "sun": "☀️", "clap": "👏", "sparkles": "✨",
  ]

  static func builtinStickerIDs() -> [String] { Array(builtinStickers.keys).sorted() }
  static func glyph(for stickerID: String) -> String { builtinStickers[stickerID] ?? "★" }

  static func render(_ base: UIImage, layers: [PhotoLayer], imageResolver: (String) -> UIImage? = { _ in nil }) -> UIImage {
    guard !layers.isEmpty else { return base }
    let renderer = PhotoEditSession.pixelRenderer(size: base.size)
    return renderer.image { context in
      base.draw(in: CGRect(origin: .zero, size: base.size))
      for layer in layers where layer.visible {
        let geometry = OverlayGeometry.map(layer, frameSize: base.size)
        let ctx = context.cgContext
        ctx.saveGState()
        ctx.translateBy(x: geometry.center.x, y: geometry.center.y)
        ctx.rotate(by: geometry.rotationDegrees * .pi / 180)
        ctx.scaleBy(x: geometry.scale, y: geometry.scale)
        switch layer.type {
        case .text: drawText(layer, shortSide: geometry.shortSide)
        case .sticker: drawSticker(layer, shortSide: geometry.shortSide, resolver: imageResolver)
        case .overlay: drawOverlay(layer, canvasWidth: base.size.width, resolver: imageResolver)
        case .drawing: drawStroke(ctx, layer, width: base.size.width, height: base.size.height, shortSide: geometry.shortSide)
        }
        ctx.restoreGState()
      }
    }
  }

  private static func drawText(_ layer: PhotoLayer, shortSide: CGFloat) {
    let fontSize = layer.fontSize / 48 * (shortSide * 0.06)
    let font: UIFont = layer.fontFamily.flatMap { UIFont(name: $0, size: fontSize) } ?? .systemFont(ofSize: fontSize)
    let paragraph = NSMutableParagraphStyle()
    paragraph.alignment = .center
    let attributes: [NSAttributedString.Key: Any] = [
      .font: font,
      .foregroundColor: layer.textColor.withAlphaComponent(layer.opacity),
      .paragraphStyle: paragraph,
    ]
    let text = layer.text.isEmpty ? " " : layer.text
    let attributed = NSAttributedString(string: text, attributes: attributes)
    let size = attributed.size()
    attributed.draw(at: CGPoint(x: -size.width / 2, y: -size.height / 2))
  }

  private static func drawSticker(_ layer: PhotoLayer, shortSide: CGFloat, resolver: (String) -> UIImage?) {
    let size = shortSide * 0.18
    let rect = CGRect(x: -size / 2, y: -size / 2, width: size, height: size)
    if let uri = layer.stickerUri, let image = resolver(uri) {
      image.draw(in: rect, blendMode: .normal, alpha: layer.opacity)
      return
    }
    let glyph = layer.stickerId.map { PhotoLayerRenderer.glyph(for: $0) } ?? "★"
    let attributes: [NSAttributedString.Key: Any] = [
      .font: UIFont.systemFont(ofSize: size),
    ]
    let attributed = NSAttributedString(string: glyph, attributes: attributes)
    let textSize = attributed.size()
    attributed.draw(at: CGPoint(x: -textSize.width / 2, y: -size / 2), withAlpha: layer.opacity)
  }

  /// Draws a user-uploaded overlay image scaled to fit a box whose width equals the full canvas
  /// width and whose height is derived from `overlayAspectRatio` — the layer's own `scale` (already
  /// applied to the drawing context by `render`) then shrinks/grows it from there. Unlike stickers,
  /// there is no glyph fallback: if the resolver can't produce an image, nothing is drawn.
  private static func drawOverlay(_ layer: PhotoLayer, canvasWidth: CGFloat, resolver: (String) -> UIImage?) {
    guard let uri = layer.overlayUri, let image = resolver(uri) else { return }
    let aspectRatio = layer.overlayAspectRatio.flatMap { $0 > 0 ? $0 : nil } ?? (image.size.width / max(image.size.height, 1))
    let width = canvasWidth
    let height = width / max(aspectRatio, 0.0001)
    let rect = CGRect(x: -width / 2, y: -height / 2, width: width, height: height)
    image.draw(in: rect, blendMode: .normal, alpha: layer.opacity)
  }

  private static func drawStroke(_ ctx: CGContext, _ layer: PhotoLayer, width: CGFloat, height: CGFloat, shortSide: CGFloat) {
    guard layer.drawPoints.count >= 2 else { return }
    ctx.setAlpha(layer.opacity)
    ctx.setStrokeColor(layer.drawColor.cgColor)
    ctx.setLineWidth(layer.drawStrokeWidth * shortSide)
    ctx.setLineCap(.round)
    ctx.setLineJoin(.round)
    let points = layer.drawPoints.map { CGPoint(x: $0.0 * width, y: $0.1 * height) }
    ctx.move(to: points[0])
    points.dropFirst().forEach { ctx.addLine(to: $0) }
    ctx.strokePath()
  }

}

private extension NSAttributedString {
  /// `draw(at:)` ignores `withAlpha` unless applied via the graphics context; this composites the whole glyph at `alpha`.
  func draw(at point: CGPoint, withAlpha alpha: CGFloat) {
    guard let ctx = UIGraphicsGetCurrentContext() else { draw(at: point); return }
    ctx.saveGState()
    ctx.setAlpha(alpha)
    draw(at: point)
    ctx.restoreGState()
  }
}
